### PR TITLE
Provide method for unlocking keyring by decrypted key

### PR DESCRIPTION
**Pull request by darkwing**
Currently just an outline for for unlocking by decrypted key for MV3

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -170,12 +171,23 @@ class KeyringController extends EventEmitter {
...
+    this.fullUpdate();
+
+    return this.encryptedKey;
+  }
+
+  /* 
+    MV3:  Unlock by encrypted key
+  */
+  async submitEncryptedKey(encryptedKey) {
+    this.keyrings = await this.unlockKeyrings(undefined, encryptedKey);
```
Comment (naugtur) [inline index.js:186]: (could) make it two methods in the API to avoid a chance of confusion - passing the correct secret into the wrong slot would result in what seems to be an incorrect secret and waste time debugging

File: index.js [hunk 2]
```diff
@@ -546,14 +559,40 @@ class KeyringController extends EventEmitter {
...
+      // MV3: Since there's a new salt, we need to generate a new encrypted key
+      // for use in the
+      this.encryptedKey = this._generateEncryptedKey(password, salt);
+    } else {
+      const encryptedVault = this.store.getState().vault;
+      if (!encryptedVault) {
+        throw new Error('Cannot unlock without a previous vault.');
+      }
+      // MV3: We can use an existing salt if one exists in the encrypted key
+      [, salt] = encryptedVault.split(VAULT_SEPARATOR);
```
Comment (naugtur) [inline index.js:593]: (could) In my (very limited) experience it pays off to hide the logic of extracting bits of string behind a tiny abstraction if you use it in more than one place. If we ever end up in a situation where we need to migrate from one separator to another or something, we're much better off having a test-covered abstraction where we can quickly introduce tests for the separator migration or any complications we have to add to the salt storage.

File: index.js [hunk 3]
```diff
@@ -585,20 +627,73 @@ class KeyringController extends EventEmitter {
...
 
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    // MV3: If the separator string is in the vault string, the user has already migrated
+    // from the previous password-only model
+    let vault = null;
+    if (encryptedVault && encryptedVault.includes(VAULT_SEPARATOR)) {
+      const [, salt] = encryptedVault.split(VAULT_SEPARATOR);
```
Comment (naugtur) [inline index.js:642]: The logic of checking what sort of vault content we're dealing with and whether it supports a key or a password would also end up in the abstraction

File: index.js [hunk 4]
```diff
@@ -585,20 +627,73 @@ class KeyringController extends EventEmitter {
...
+      await this.persistAllKeyrings(password);
+    }
+
     return this.keyrings;
   }
 
+  // MV3:  Generates the encrypted key
+  _generateEncryptedKey(password, salt) {
+    // TODO: How complicaated do we want this to be?
+    return window.btoa(password + salt);
```
Comment (naugtur) [inline index.js:675]: (must) We want it to be a hash of the two, so that plaintext password is not recoverable. 
(could) Starting with the salt might be preferred for some hashing functions IIRC

File: index.js [hunk 5]
```diff
@@ -585,20 +627,73 @@ class KeyringController extends EventEmitter {
...
+  // https://github.com/MetaMask/browser-passworder/blob/d24b66934ab4d39d97ccff9ff0c3d25f86f1a141/src/index.ts#L180
+  _generateSalt(byteCount = 32) {
+    const view = new Uint8Array(byteCount);
+    global.crypto.getRandomValues(view);
+    // Uint8Array is a fixed length array and thus does not have methods like pop, etc
+    // so TypeScript complains about casting it to an array. Array.from() works here for
+    // getting the proper type, but it results in a functional difference. In order to
+    // cast, you have to first cast view to unknown then cast the unknown value to number[]
+    // TypeScript ftw: double opt in to write potentially type-mismatched code.
+    const b64encoded = window.btoa(String.fromCharCode.apply(null, view));
```
Comment (naugtur) [inline index.js:693]: It's a random salt, so maybe you can skip all the pain and use `crypto.randomUUID()` instead.
I don't know what the requirements for entrpoy are though. UUID is 122 bits. I'm not qualified to say if that's enough ;)

Comment (Gudahtt) [inline index.js:693]: We could use `Buffer.from(...).toString('base64')` here instead of `window.btoa`. That would ensure this library still works in a Node.js context.

Comment (Gudahtt) [reply index.js:693]: Oh, i see this was taken from `browser-passworder`, which is only meant to be used by browsers.

Any reason this was inlined, rather than calling `this.encryptor.generateSalt`?

File: index.js [hunk 6]
```diff
@@ -546,14 +559,40 @@ class KeyringController extends EventEmitter {
...
+  persistAllKeyrings(password) {
+    if (password && typeof password !== 'string') {
       return Promise.reject(
         new Error('KeyringController - password is not a string'),
       );
     }
 
-    this.password = password;
+    // MV3: Since we also allow persisting without a password, we should require this.encryptedKey
+    if (password === undefined && this.encryptedKey === undefined) {
```
Comment (darkwing) [inline index.js:570]: All of our tests are failing on this check, which is problematic because `persistAllKeyrings` is called without a password and without encryption key, from a few different places.

File: index.js (hunk 1 above)
Comment (darkwing) [reply index.js:186]: I thought  about this but that method is kinda complicated and so duplicating it would be a maintenance problem.

File: index.js (hunk 2 above)
Comment (darkwing) [reply index.js:594]: Done!

File: index.js (hunk 6 above)
Comment (danjm) [reply index.js:571]: I believe the problem is that before the changes in this PR, the tests relied on the `this.password = password;` of what used to be line 522 in the `createFirstKeyTree` function

The stack trace for the error in the tests is:
```
  at KeyringController.persistAllKeyrings (index.js:573:9)
      at index.js:239:21
      at Object.<anonymous> (test/index.js:31:5)
```

So it is happening after calling `await keyringController.createNewVaultAndKeychain(password);` on line 31 in the test file, which is in the `beforeEach` function. `createNewVaultAndKeychain` calls `this.createFirstKeyTree(password)`, but passing the password to `createFirstKeyTree` no longer does anything. So...

I think the fix is to pass the password from `createFirstKeyTree` to `this.addNewKeyring` to `this.persistAllKeyrings()`

File: index.js [hunk 7]
```diff
@@ -516,12 +537,39 @@ class KeyringController extends EventEmitter {
...
+
+    let salt = null;
+    if (password) {
+      // MV3: If this is a migration or new password-driven login, we should
+      // create or rotate the salt
+      salt = encryptor.generateSalt();
+
+      // MV3: Since there's a new salt, we need to generate a new encrypted key
+      // for use in the
+      this.encryptedKey = await this._generateEncryptedKey(password, salt);
```
Comment (darkwing) [inline index.js:563]: `this. _generateEncryptedKey` returns an `ArrayBuffer`, so we'll need to convert this to string.

File: index.js [hunk 8]
```diff
@@ -548,20 +602,72 @@ class KeyringController extends EventEmitter {
...
+  parseVault(encryptedVault) {
+    const [vault, salt] = encryptedVault.split(VAULT_SEPARATOR);
+    return { vault, salt };
+  }
+
+  // MV3:  Generates the encrypted key
+  async _generateEncryptedKey(password, salt) {
+    const data = new TextEncoder("utf-8").encode(password + salt);
+    const encryptedSha = await crypto.subtle.digest('SHA-256', data);
+    const encryptedKey = new TextDecoder("utf-8").decode(encryptedSha);
```
Comment (darkwing) [inline index.js:662]: The result of this function is generating this error:

```
Unexpected non-whitespace character after JSON at position 1425
```

An example encrypted key is `'nhd\x10x1.ۇ\x04bh$K��{ZT3!L\x06�B����"��'`

File: test/index.js [hunk 9]
```diff
@@ -20,6 +22,20 @@ const walletTwoAddresses = [
...
+
+// Shims for globals that don't exist in test environment
+global.crypto = {
+  getRandomValues: () => {
+    return Math.floor(Math.random(1000000));
+  },
+  subtle: {
+    digest: (algorithm, data) =>
+      new Uint8Array(Date.now().toString().split('')).buffer,
+  },
```
Comment (darkwing) [inline test/index.js:34]: At present, a few of the tests are being effected by the fact that I believe `TextEncoder.decode` is working improperly in the tests.  Each time it appears to return an empty string, which it shouldn't be.  So either the shim is off or the result of what I'm shimming `digest` to do is wrong.

File: test/index.js [hunk 10]
```diff
@@ -451,4 +473,79 @@ describe('KeyringController', function () {
...
+
+      // Log the user out
+      await keyringController.setLocked();
+
+      // Attempt to login with an incorrect encryption key
+      const encryptionKey = await keyringController.submitEncryptedKey(
+        'FAKE KEY',
+      );
+
+      expect(encryptionKey).rejects.toThrow();
```
Comment (darkwing) [inline test/index.js:513]: This test is still passing, which it shouldn't.  I think this is affected by the `_generateEncryptedKey` returning what looks like an empty string in the just the tests.

Comment (naugtur) [reply test/index.js:513]: There's no way this test could work. 
You're using an async assertion which should be getting a promise as input, but you await that promise above and pass the return value.

Avoid awaiting the  submitEncryptedKey call, assign the promise it returns to a variable and/or put that reference inside expect(_).rejects

File: test/index.js [hunk 11]
```diff
@@ -451,4 +473,79 @@ describe('KeyringController', function () {
...
+
+      const currentKey = keyringController.getEncryptedKey();
+
+      await keyringController.setLocked();
+
+      await keyringController.submitPassword(password);
+
+      const newKey = keyringController.getEncryptedKey();
+
+      expect(currentKey === newKey).toBe(false);
```
Comment (naugtur) [inline test/index.js:548]: if you use a "not equal" assertion here, you'll get both values in test failure output when it fails, instead of "expected true to be false"

File: test/index.js [hunk 12]
```diff
@@ -451,4 +473,79 @@ describe('KeyringController', function () {
...
+
+      const encryptionKey = keyringController.getEncryptedKey();
+
+      // Log the user out
+      await keyringController.setLocked();
+
+      // Attempt to login with an incorrect encryption key
+      const key = await keyringController.submitEncryptedKey(encryptionKey);
+
+      expect(key).toBeTruthy();
```
Comment (naugtur) [inline test/index.js:527]: Is it in any way important (or relevant) that key is truthy?
Seems like this should assert  submitEncryptedKey() doesn't reject the promise

File: index.js [hunk 13]
```diff
@@ -171,6 +181,19 @@ class KeyringController extends EventEmitter {
...
+
+  /* 
+    MV3:  Unlock by encrypted key
+  */
+  async submitEncryptedKey(encryptedKey) {
+    this.keyrings = await this.unlockKeyrings(undefined, encryptedKey);
+    this.setUnlocked();
+    this.fullUpdate();
+
+    return this.encryptedKey;
```
Comment (naugtur) [inline index.js:196]: What's the purpose of this being returned? Is it only for tests? 

If so, there's nothing stopping you from making an assertion on `theReferenceTheTestHas.encryptedKey` instead. It's not a private field.

File: index.js [hunk 14]
```diff
@@ -186,7 +209,9 @@ class KeyringController extends EventEmitter {
     if (!encryptedVault) {
       throw new Error('Cannot unlock without a previous vault.');
     }
-    await this.encryptor.decrypt(password, encryptedVault);
+
+    const result = await this.encryptor.decrypt(password, encryptedVault);
```
Comment (naugtur) [inline index.js:213]: (nit) the more common spelling seems to be encrypter :shrug:  
https://en.wiktionary.org/wiki/encryptor

File: index.js [hunk 15]
```diff
@@ -186,7 +209,9 @@ class KeyringController extends EventEmitter {
...
+    return result;
```
Comment (naugtur) [inline index.js:214]: jsdoc doesn't suggest this would return anything. Why is that needed?

File: index.js [hunk 16]
```diff
@@ -516,12 +540,39 @@ class KeyringController extends EventEmitter {
...
-    this.password = password;
+    // MV3: Since we also allow persisting without a password,
+    // we should require this.encryptedKey
+    if (password === undefined && this.encryptedKey === undefined) {
+      return Promise.reject(
+        new Error(
+          'KeyringController - a password or encryptedKey must exist to persist keyrings',
+        ),
+      );
+    }
```
Comment (naugtur) [inline index.js:556]: Throwing inside an async function has the same result of rejecting the promise returned by it. 
Is this an optimization? Was throwing noticably slower?

File: index.js [hunk 17]
```diff
@@ -532,10 +583,16 @@ class KeyringController extends EventEmitter {
       }),
     );
     const encryptedString = await this.encryptor.encrypt(
-      this.password,
+      this.encryptedKey,
       serializedKeyrings,
     );
-    this.store.updateState({ vault: encryptedString });
+
+    const newVault = [encryptedString, VAULT_SEPARATOR, salt].join('');
```
Comment (naugtur) [inline index.js:590]: using a template string will sometimes throw

Comment (naugtur) [inline index.js:590]: (could) assert both values are not undefined 
If we introduce a bug somewhere in the future it'd be better to throw an error and leave the password encrypted vault as is than to save something like `VAULT:::undefined` accidentally

File: index.js [hunk 18]
```diff
@@ -548,20 +605,72 @@ class KeyringController extends EventEmitter {
...
 
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    // MV3: If the separator string is in the vault string, the user has already migrated
+    // from the previous password-only model
+    let vault = null;
+    if (encryptedVault?.includes?.(VAULT_SEPARATOR)) {
+      const { salt, vault: vaultOnly } = this.parseVault(encryptedVault);
```
Comment (naugtur) [inline index.js:620]: (nit)
since this codebase is OO style more than it is FP, maybe a simple Vault class with a constructor that can create the vault object by parsing the string or getting both chunks as input would look nicer. Something like the global URL() which can be used to parse URLs or create them from pieces.

It'd expose public fields:
- salt
- vault
- serialized (the string resulting from concatenation or passed in for parsing)
- isSalted (to use instead of `encryptedVault?.includes?.(VAULT_SEPARATOR)` here)

SEPARATOR would not be known outside the Vault class

File: index.js [hunk 19]
```diff
@@ -548,20 +605,72 @@ class KeyringController extends EventEmitter {
...
+
+        log('[unlockKeyrings] *new* encryptedKey is: ', this.encryptedKey);
+
+      } else if (encryptedKey !== undefined) {
+        log('[unlockKeyrings] *old* encryptedKey is: ', this.encryptedKey);
+        this.encryptedKey = encryptedKey;
+      } else {
+        throw new Error(
+          'No way to decrypt a salted vault without a password or encrypted key',
+        );
```
Comment (naugtur) [inline index.js:635]: (could) move the error to the top, throw early if both are missing.

File: index.js [hunk 20]
```diff
@@ -548,20 +605,72 @@ class KeyringController extends EventEmitter {
    * @param {string} password - The keyring controller password.
    * @returns {Promise<Array<Keyring>>} The keyrings.
    */
-  async unlockKeyrings(password) {
+  async unlockKeyrings(password, encryptedKey) {
```
Comment (naugtur) [inline index.js:608]: :thinking: Why is it called encryptedKey not encryptionKey? It's not like the key itself is encrypted.

File: index.js [hunk 21]
```diff
@@ -548,20 +605,72 @@ class KeyringController extends EventEmitter {
...
+    } else {
+      vault = await this.encryptor.decrypt(password, encryptedVault);
+    }
+
     await Promise.all(vault.map(this._restoreKeyring.bind(this)));
+
     await this._updateMemStoreKeyrings();
+
+    // MV3: If we're provided a password, we should persist keyrings
+    // so that we can either (1) migrate or (2) create a new salt
```
Comment (danjm) [inline index.js:648]: I think the below if block should probably go in `submitPassword`, after `unlockKeyrings` resolves

File: test/index.js [hunk 22]
```diff
@@ -20,6 +22,20 @@ const walletTwoAddresses = [
   '0xbbafcf3d00fb625b65bb1497c94bf42c1f4b3e78',
   '0x49dd2653f38f75d40fdbd51e83b9c9724c87f7eb',
 ];
+
+// Shims for globals that don't exist in test environment
+global.crypto = {
```
Comment (naugtur) [inline test/index.js:27]: FYI, available in node 15
https://nodejs.org/api/webcrypto.html#web-crypto-api

File: test/index.js [hunk 23]
```diff
@@ -20,6 +22,20 @@ const walletTwoAddresses = [
...
+  getRandomValues: () => {
+    return Math.floor(Math.random(1000000));
+  },
+  subtle: {
+    digest: (algorithm, data) =>
+      new Uint8Array(Date.now().toString().split('')).buffer,
+  },
+};
```
Comment (naugtur) [inline test/index.js:35]: the shim could use existing Node crypto implementation if you want results that look more like the originals

```suggestion
const nodeCrypto = require('crypto');
global.crypto = {
  getRandomValues: () => {
    return nodeCrypto.randomFillSync(buffer);
  },
  subtle: {
    digest: (algorithm, data) => {
      const hash = nodeCrypto.createHash(algorithm);
      return hash.update(data).digest();
  },
};
```
^ warning: may contain typos

File: test/index.js [hunk 24]
```diff
@@ -451,4 +473,79 @@ describe('KeyringController', function () {
...
     });
   });
+
+  describe('Manifest v3 changes', function () {
+    it('can still verify password', async function () {
+      // Log the user out
+      await keyringController.setLocked();
+
+      // Attempt to verify the password
+      const result = await keyringController.verifyPassword(password);
```
Comment (naugtur) [inline test/index.js:483]: (nit) verifyPassword throws on failure, so the assertion is not reached if the test won't pass.
using async assertions may produce more helpful output from Jest
```suggestion
     await expect(keyringController.verifyPassword(password)).resolves.toBeTruthy()
```

File: test/index.js [hunk 25]
```diff
@@ -451,4 +473,79 @@ describe('KeyringController', function () {
...
+      expect(result).toBeTruthy();
+    });
+
+    it('can still unlock with password after being migrated and locked', async function () {
```
Comment (naugtur) [inline test/index.js:487]: It's unclear to me. Where did the migration happen?
A `before` call would help me understand what's the precondition, even if redundant.

File: test/index.js (hunk 23 above)
Comment (naugtur) [reply test/index.js:35]: OR
this could be stubbed with sinon in a before or beforeEach and random values would no longer be random so they could be used in assertions

File: test/index.js [hunk 26]
```diff
@@ -451,4 +473,79 @@ describe('KeyringController', function () {
...
+      // Log the user out
+      await keyringController.setLocked();
+
+      // Attempt to unlock the keychaing via old password
+      const result = await keyringController.submitPassword(password);
+      expect(result).toBeTruthy();
+
+      // Ensure the new vault value has separator and salt
+      const { vault, salt } = keyringController.store.getState();
+      expect(vault).toContain(':::');
```
Comment (naugtur) [inline test/index.js:497]: 1. testing an implementation detail. 
2. How is ':::' present in the vault field while salt is already available separately?

If you can use a deterministic shim for getRandomValues you could assert that the salt is correct and present where necessary.

File: test/index.js (hunk 10 above)
Comment (naugtur) [inline test/index.js:513]:
```suggestion
      const controllerPromise = keyringController.submitEncryptedKey(
        'FAKE KEY',
      );

      await expect(controllerPromise).rejects.toThrow();
```

File: index.js [hunk 27]
```diff
@@ -171,6 +181,19 @@ class KeyringController extends EventEmitter {
...
+    return this.encryptedKey;
+  }
+
+  /* 
+    MV3:  Unlock by encrypted key
+  */
+  async submitEncryptedKey(encryptedKey) {
+    this.keyrings = await this.unlockKeyrings(undefined, encryptedKey);
+    this.setUnlocked();
+    this.fullUpdate();
```
Comment (naugtur) [inline index.js:194]: I don't know what this part is doing, assuming some signaling that's not relevant to the security or lock/unlock process.

File: index.js [hunk 28]
```diff
@@ -548,20 +605,72 @@ class KeyringController extends EventEmitter {
...
+
+      if (password !== undefined) {
+        this.encryptedKey = await this._generateEncryptedKey(password, salt);
+
+        log('[unlockKeyrings] *new* encryptedKey is: ', this.encryptedKey);
+
+      } else if (encryptedKey !== undefined) {
+        log('[unlockKeyrings] *old* encryptedKey is: ', this.encryptedKey);
+        this.encryptedKey = encryptedKey;
+      } else {
```
Comment (naugtur) [inline index.js:632]: This logic is good :+1:

File: index.js [hunk 29]
```diff
@@ -548,20 +605,72 @@ class KeyringController extends EventEmitter {
...
+      const { salt, vault: vaultOnly } = this.parseVault(encryptedVault);
+
+      log('[unlockKeyrings] salt is: ', salt, '; encryptedVault is:', encryptedVault);
+
+      if (password !== undefined) {
+        this.encryptedKey = await this._generateEncryptedKey(password, salt);
+
+        log('[unlockKeyrings] *new* encryptedKey is: ', this.encryptedKey);
+
+      } else if (encryptedKey !== undefined) {
```
Comment (naugtur) [inline index.js:629]: This logic is good :+1:

File: index.js [hunk 30]
```diff
@@ -516,12 +540,39 @@ class KeyringController extends EventEmitter {
...
+
+    let salt = null;
+    if (password) {
+      // MV3: If this is a migration or new password-driven login, we should
+      // create or rotate the salt
+      salt = encryptor.generateSalt();
+
+      // MV3: Since there's a new salt, we need to generate a new encrypted key
+      // for use in the
+      this.encryptedKey = await this._generateEncryptedKey(password, salt);
```
Comment (naugtur) [inline index.js:566]: We're overriding the encryptedKey field before going through the (**asynchronous**) process of preparing the new vault.
That leaves a possibility (because of bad timing or an error) for the key in this.encryptedKey to not be a key that can open the current vault.

It's not a big deal, worst case scenario it fails and forces the user to supply password again. Might be annoying if happening right after they supplied the password already.

Keeping the key in local variable and setting it in the same synchronous execution as the call to store the new vault would eliminate that gap.


And while we're at it - where/when is the key being saved to survive a restart? Will this flow save it too early? Why is it not saved in the same synchronous execution as the vault it unlocks?

File: test/index.js (hunk 11 above)
Comment (darkwing) [reply test/index.js:548]: Good improvement!

File: index.js (hunk 13 above)
Comment (darkwing) [reply index.js:196]: That's fair.

File: index.js (hunk 14 above)
Comment (darkwing) [reply index.js:213]: Agreed, but this is an existing practice that doesn't need to be changed in this PR.

File: index.js (hunk 15 above)
Comment (darkwing) [reply index.js:214]: This returns:

```
[
      {
        type: 'HD Key Tree',
        data: {
          mnemonic: [Array],
          numberOfAccounts: 1,
          hdPath: "m/44'/60'/0'/0"
        }
      }
    ]
```

I added for debugging purposes, but yes, it can be removed.

File: index.js (hunk 17 above)
Comment (darkwing) [reply index.js:590]: Good idea!

File: test/index.js (hunk 23 above)
Comment (darkwing) [reply test/index.js:35]: Getting error `Digest method not supported` :/

File: index.js [hunk 31]
```diff
@@ -548,20 +637,50 @@ class KeyringController extends EventEmitter {
...
   }
 
+  parseVault(encryptedVault) {
+    const [vault, salt] = encryptedVault.split(VAULT_SEPARATOR);
+    return { vault, salt };
+  }
+
+  // MV3:  Generates the encrypted key
+  async _generateEncryptionKey(password, salt) {
+    return toHex(sha256(utf8ToBytes(password + salt)));
```
Comment (Gudahtt) [inline index.js:676]: Why are we using `ethereum-cryptography/sha256` here instead of the encryptor?

File: index.js [hunk 32]
```diff
@@ -168,7 +176,24 @@ class KeyringController extends EventEmitter {
    * @returns {Promise<Object>} A Promise that resolves to the state.
    */
   async submitPassword(password) {
+    await this.verifyPassword(password);
     this.keyrings = await this.unlockKeyrings(password);
+
+    // MV3: If we're provided a password, we should persist keyrings
```
Comment (Gudahtt) [inline index.js:182]: This package is used by mobile and MV2 extension as well, and these changes to the workflow will apply to both. This "MV3:" prefix might be confusing or misleading.

File: index.js (hunk 31 above)
Comment (danjm) [reply index.js:697]: @darkwing I know this code came from the branch I pushed up, but you might have missed this comment in our DM with Zb:

> I think the shim could be used for stubbing random salt with predictable values and asserting for it.
Also, getting a dependency instead using a feature in the browser is not only a detriment to maintainability but also performance :shushing_face:

So we should go back to your old implementation that uses the native crypto api directly. But as per Mark's comment, we can probably just use the encryptor itself

File: index.js [hunk 33]
```diff
@@ -168,7 +174,29 @@ class KeyringController extends EventEmitter {
    * @returns {Promise<Object>} A Promise that resolves to the state.
    */
   async submitPassword(password) {
+    await this.verifyPassword(password);
     this.keyrings = await this.unlockKeyrings(password);
+
+    // If we're provided a password, we should persist keyrings
```
Comment (danjm) [inline index.js:180]: Nit: this comment should be updated because a password is not optional at this point. I might reword this as: `persistAllKeyrings` is here called so that we can either (1) migrate or (2) create a new salt

File: index.js (hunk 31 above)
Comment (danjm) [reply index.js:697]: to be more specific, you could use `keyFromPassword(password, salt)`

Comment (danjm) [reply index.js:697]: But, we might want to consider how we can more fully leverage browser passworder and push more logic down to that

File: index.js [hunk 34]
```diff
@@ -548,20 +641,62 @@ class KeyringController extends EventEmitter {
...
+      throw new Error(
+        'No way to decrypt a salted vault without a password or encrypted key',
+      );
+    }
+
     await this.clearKeyrings();
-    const vault = await this.encryptor.decrypt(password, encryptedVault);
-    this.password = password;
+
+    // If the separator string is in the vault string, the user has already migrated
```
Comment (danjm) [inline index.js:658]: I think this comment is supposed to be on line 620?

File: index.js (hunk 31 above)
Comment (darkwing) [reply index.js:698]: I tried doing the following:

```
async _generateEncryptionKey(password, salt) {
    //return toHex(sha256(utf8ToBytes(password + salt)));
    const key = await encryptor.keyFromPassword(password, salt);
    const exportedKey = await window.crypto.subtle.exportKey('jwk', key);
    const keyString = new Uint8Array(exportedKey).toString();

    console.log("key is: ", key, exportedKey);
    console.log("keystring is: ", keyString);
    return key;
  }
```

Result being:

AttemptedKeyExport

The string form of extracted key is empty.  I don't think this helps us.

Comment (Gudahtt) [reply index.js:698]: It's not totally clear what you're trying to do in that snippet, or how it relates to this thread. But that is failing because you're asking for the key to be exported as a JWT, which is an object. Then you're passing the object into the Uint8Array constructor, which isn't meant to accept that sort of object as input. It tries to iterate over it but it's not iterable, giving you an empty arraybuffer, thus the empty string.
